tools/codegen/model_generator/model.go

```diff
@@ -0,0 +1,257 @@
+		return *model, err
+	}
+
+	// imports only referenced by the Constructor message.
+	model.addImports(parser, templateProto, cnstrDesc)
+	if err != nil {
+		return *model, err
+	}
+
+	return *model, err
+}
+
+// Find the file that has the options TemplateVariety and TemplateName. There should only be one such file.
+func getTmplFileDesc(fds []*FileDescriptor) (*FileDescriptor, error) {
+	var templateDescriptorProto *FileDescriptor = nil
+
+	for _, fd := range fds {
+		if fd.GetOptions() == nil {
+			continue
+		}
+		if !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) && !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
+			continue
+		} else if proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) && proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
+			if templateDescriptorProto == nil {
+				templateDescriptorProto = fd
+			} else {
+				return nil, fmt.Errorf("Proto files %s and %s, both have"+
+					" the options %s and %s. Only one proto file is allowed with those options",
+					fd.GetName(), templateDescriptorProto.Name,
+					tmplExtns.E_TemplateVariety.Name, tmplExtns.E_TemplateName.Name)
+
+			}
+		} else {
+			return nil, fmt.Errorf("Proto files %s has only one of the "+
+				"following two options %s and %s. Both options are required.",
+				fd.GetName(), tmplExtns.E_TemplateVariety.Name, tmplExtns.E_TemplateName.Name)
+		}
+	}
+	if templateDescriptorProto == nil {
+		return nil, fmt.Errorf("There has to be one proto file that has both extensions %s and %s",
+			tmplExtns.E_TemplateVariety.Name, tmplExtns.E_TemplateVariety.Name)
+	}
+	return templateDescriptorProto, nil
+}
+
+// Add all the file level fields to the model.
+func (m *Model) addTopLevelFields(fd *FileDescriptor) error {
+	if fd.Package == nil {
+		return fmt.Errorf("package name missing on file %s", fd.GetName())
+	}
+	m.PackageName = goPackageName(strings.TrimSpace(*fd.Package))
+
+	tmplName, err := proto.GetExtension(fd.GetOptions(), tmplExtns.E_TemplateName)
+	if err != nil {
+		// This func should only get called for FileDescriptor that has this attribute,
+		// therefore it is impossible to get to this state.
+		return fmt.Errorf("file option %s is required", tmplExtns.E_TemplateName.Name)
+	}
+	if name, ok := tmplName.(*string); !ok {
+		// protoc should mandate the type. It is impossible to get to this state.
+		return fmt.Errorf("%s should be of type string", tmplExtns.E_TemplateName.Name)
+	} else {
+		m.Name = *name
+	}
+
+	tmplVariety, err := proto.GetExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety)
+	if err != nil {
+		// This func should only get called for FileDescriptor that has this attribute,
+		// therefore it is impossible to get to this state.
+		return fmt.Errorf("file option %s is required", tmplExtns.E_TemplateVariety.Name)
+	}
+	if *(tmplVariety.(*tmplExtns.TemplateVariety)) == tmplExtns.TemplateVariety_TEMPLATE_VARIETY_CHECK {
+		m.Check = true
+		m.VarietyName = "Check"
+	} else {
```

Comment:
Should test for _TEMPLATE_VARIETY_REPORT explicitly, and fail for any other value. Right now, this would take any unknown value as REPORT which isn't right.